### DISCUSSION TITLE
Using FastAPI OAuth scope to replace user role based system.

### DISCUSSION THREAD

**Discussion by louprogramming** [fastapi/fastapi#8965] [Questions]
### First check

* [*] I added a very descriptive title to this issue.
* [*] I used the GitHub search to find a similar issue and didn't find it.
* [*] I searched the FastAPI documentation, with the integrated search.
* [*] I already searched in Google "How to X in FastAPI" and didn't find any information.
* [*] I already read and followed all the tutorial in the docs and didn't find an answer.
* [*] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
* [*] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
* [*] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).
* [*] After submitting this, I commit to one of:
    * Read open issues with questions until I find 2 issues where I can help someone and add a comment to help there.
    * I already hit the "watch" button in this repository to receive notifications and I commit to help at least 2 people that ask questions in the future.
    * Implement a Pull Request for a confirmed bug.

### Example

I have a system which has role based access system.

When the API is called, first the user is authenticated, then I further verify if the user has an access to the endpoint before actually calling the endpoint.

Here is an example for the roles I have:

```
WRITE = {"all": "write_access"} # this is used in user access verification method
READ = {"all": "read_access"} # this is used in user access verification method
    
ACCESS = {
    # Administrator has every access to every endpoint.
    "Administrator": WRITE,
    
    # Basic User
    # Restricted access to endpoints
    # Can switch ownership.
    # e.g. A basic user has only READ access to endpoint_1.
    "Basic": {
        "endpoint_1": READ,
        "endpoint_2": WRITE,
        ...
        "endpoint_n": READ
    },
    "Premium": {
    ... 
    }


```
### Description

I was reading the FastAPI documentation here: https://fastapi.tiangolo.com/advanced/security/oauth2-scopes/. FastAPI scopes allows more fine-grained permission system, that restricts access to the endpoints. It seems like the way it works is we give scopes directly to each endpoint. Thus when API calls hit the endpoint, we check if the calling user has correct scope to access the endpoint. I believe this is a bit different from how I'm currently handling the user access verification, where instead of giving scope to each method, I attach the accesses directly to each user type and have a custom method dedicated to check the user verification. In this case, would FastAPI Oauth2 scope be a good choice for my purpose? I'm currently not using FastAPI security or dependencies for authentication of JWT, but have my own internal authentication workflow. I would like to set up a more fine-grained permission system and as I'm moving my system to FastAPI, it would be helpful if I could benefit from built-in functionality which is scope in this case. Any help would be greatly appreciated!

**Comment by YuriiMotov** [comment]
You should try it and decide
